### PR TITLE
Support for extra lazy get for both owning and inverse side on many to many associations.

### PR DESCRIPTION
**Pull request by guilhermeblanco**
Addresses #710 and #712

### CODE REVIEW

File: lib/Doctrine/ORM/Persisters/ManyToManyPersister.php [hunk 1]
```diff
@@ -77,7 +77,19 @@ public function update(PersistentCollection $coll)
      */
     public function get(PersistentCollection $coll, $index)
     {
-        throw new \BadMethodCallException("Selecting a collection by index is not supported by this CollectionPersister.");
+        $mapping = $coll->getMapping();
+
+        if ( ! isset($mapping['indexBy'])) {
+            throw new \BadMethodCallException("Selecting a collection by index is only supported on indexed collections.");
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/ManyToManyPersister.php]: Specific exceptions please

File: tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php [hunk 2]
```diff
@@ -973,11 +1012,13 @@ private function loadFixture()
...
         $this->userId2 = $user2->getId();
         $this->groupId = $group1->id;
+        $this->ddc2504OtherClassId = $otherClass->id;
+        $this->ddc2504ChildClassId = $childClass1->id;
 
+        $this->username = $user1->username;
+        $this->groupname = $group1->name;
         $this->topic = $article1->topic;
         $this->phonenumber = $phonenumber1->phonenumber;
-        $this->ddc2504OtherClassId = $otherClass->id;
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php]: Any specific reason why the order was changed?

File: lib/Doctrine/ORM/Persisters/BasicEntityPersister.php [hunk 3]
```diff
@@ -1621,14 +1621,29 @@ protected function getSelectConditionStatementColumnSQL($field, $assoc = null)
         }
 
         if (isset($this->class->associationMappings[$field])) {
+            $association = $this->class->associationMappings[$field];
 
-            if ( ! $this->class->associationMappings[$field]['isOwningSide']) {
+            // Many-To-Many requires join table check for joinColumn
+            if ($association['type'] === ClassMetadata::MANY_TO_MANY) {
+                if ( ! $association['isOwningSide']) {
+                    $association = $assoc;
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/BasicEntityPersister.php]: `$association = $assoc`... can we change naming here somehow?

File: lib/Doctrine/ORM/Persisters/BasicEntityPersister.php [hunk 4]
```diff
@@ -1621,14 +1621,29 @@ protected function getSelectConditionStatementColumnSQL($field, $assoc = null)
...
+                }
+
+                $joinTableName  = $this->quoteStrategy->getJoinTableName($association, $this->class, $this->platform);
+                $joinColumn     = $assoc['isOwningSide']
+                    ? $association['joinTable']['joinColumns'][0]
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/BasicEntityPersister.php]: What if there are more than one `joinColumns`?

File: lib/Doctrine/ORM/Persisters/BasicEntityPersister.php [hunk 5]
```diff
@@ -1621,14 +1621,29 @@ protected function getSelectConditionStatementColumnSQL($field, $assoc = null)
...
+            }
+
+            if ( ! $association['isOwningSide']) {
                 throw ORMException::invalidFindByInverseAssociation($this->class->name, $field);
             }
 
-            $joinColumn = $this->class->associationMappings[$field]['joinColumns'][0];
-            $className  = (isset($this->class->associationMappings[$field]['inherited']))
-                ? $this->class->associationMappings[$field]['inherited']
+            $joinColumn = $association['joinColumns'][0];
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/BasicEntityPersister.php]: Same as above

File: lib/Doctrine/ORM/Persisters/ManyToManyPersister.php [hunk 6]
```diff
@@ -77,7 +77,19 @@ public function update(PersistentCollection $coll)
...
+        }
+
+        $persister = $this->uow->getEntityPersister($mapping['targetEntity']);
+
+        if ( ! $mapping['isOwningSide']) {
+            return $persister->load(array($mapping['mappedBy'] => $coll->getOwner(), $mapping['indexBy'] => $index), null, $mapping, array(), 0, 1);
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/ManyToManyPersister.php]: Can you avoid these one-liners? They are super-unreadable

File: lib/Doctrine/ORM/Persisters/ManyToManyPersister.php [hunk 7]
```diff
@@ -138,6 +150,12 @@ public function slice(PersistentCollection $coll, $offset, $length = null)
      */
     public function containsKey(PersistentCollection $coll, $key)
     {
+        $mapping = $coll->getMapping();
+
+        if ( ! isset($mapping['indexBy'])) {
+            throw new \BadMethodCallException("Selecting a collection by index is only supported on indexed collections.");
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/ManyToManyPersister.php]: Specific exception required

File: lib/Doctrine/ORM/Persisters/OneToManyPersister.php [hunk 8]
```diff
@@ -60,24 +60,15 @@ public function update(PersistentCollection $coll)
...
-            ),
-            null,
-            null,
-            array(),
-            null,
-            1
-        );
+        $persister = $this->uow->getEntityPersister($mapping['targetEntity']);
+
+        return $persister->load(array($mapping['mappedBy'] => $coll->getOwner(), $mapping['indexBy'] => $index), null, $mapping, array(), null, 1);
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/OneToManyPersister.php]: Avoid one-liner here

File: lib/Doctrine/ORM/Persisters/OneToManyPersister.php [hunk 9]
```diff
@@ -60,24 +60,15 @@ public function update(PersistentCollection $coll)
      */
     public function get(PersistentCollection $coll, $index)
     {
-        $mapping   = $coll->getMapping();
-        $persister = $this->uow->getEntityPersister($mapping['targetEntity']);
+        $mapping = $coll->getMapping();
 
-        if (!isset($mapping['indexBy'])) {
+        if ( ! isset($mapping['indexBy'])) {
             throw new \BadMethodCallException("Selecting a collection by index is only supported on indexed collections.");
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Persisters/OneToManyPersister.php]: Specific exception required

File: tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php [hunk 10]
```diff
@@ -757,10 +761,45 @@ public function testGetIndexByOneToMany()
     /**
      * @group DDC-1398
      */
+    public function testGetIndexByManyToManyInverseSide()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php]: Can you expand the method name for clarity?

File: tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php [hunk 11]
```diff
@@ -757,10 +761,45 @@ public function testGetIndexByOneToMany()
...
+
+        $this->assertFalse($group->users->isInitialized());
+        $this->assertEquals($queryCount + 1, $this->getCurrentQueryCount());
+        $this->assertSame($user, $this->_em->find('Doctrine\Tests\Models\CMS\CmsUser', $this->userId));
+    }
+
+    /**
+     * @group DDC-1398
+     */
+    public function testGetIndexByManyToManyOwningSide()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php]: Same as above

File: lib/Doctrine/ORM/Persisters/ManyToManyPersister.php (hunk 1 above)
Comment (guilhermeblanco) [inline lib/Doctrine/ORM/Persisters/ManyToManyPersister.php]: Not part of this scope... others does the same, keeping the standard.
Dedicated Exception should be a separate PR changing everything.

File: tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php (hunk 2 above)
Comment (guilhermeblanco) [inline tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php]: Not really... it's just that others were purely IDs... and at this block it was only objects, that's why I moved them together

File: lib/Doctrine/ORM/Persisters/BasicEntityPersister.php (hunk 3 above)
Comment (guilhermeblanco) [inline lib/Doctrine/ORM/Persisters/BasicEntityPersister.php]: Next PR will change all this.

File: lib/Doctrine/ORM/Persisters/BasicEntityPersister.php (hunk 4 above)
Comment (guilhermeblanco) [inline lib/Doctrine/ORM/Persisters/BasicEntityPersister.php]: Same as before... not supported/it would break

File: lib/Doctrine/ORM/Persisters/ManyToManyPersister.php (hunk 6 above)
Comment (guilhermeblanco) [inline lib/Doctrine/ORM/Persisters/ManyToManyPersister.php]: Will be fixed by next PR

File: tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php (hunk 10 above)
Comment (guilhermeblanco) [inline tests/Doctrine/Tests/ORM/Functional/ExtraLazyCollectionTest.php]: I just followed the standard on this class
